File: Examples/CSharp/ProgrammersGuide/Working-Shapes/CalculateCenterOfSubShapes.cs
using Aspose.Diagram;
using System;
using System.Collections.Generic;
using System.Drawing.Drawing2D;
using System.Linq;
using System.Text;

namespace CSharp.ProgrammersGuide.Working_Shapes
{
    public class CalculateCenterOfSubShapes
    {
        public static void Run()
        {
            // ExStart:CalculateCenterOfSubShapes
            // The path to the documents directory.
            string dataDir = RunExamples.GetDataDir_Shapes();

            // load Visio diagram
            Diagram diagram = new Diagram(dataDir + "Drawing1.vsdx");
            // get a group shape by ID and page index is 0
            Shape shape = diagram.Pages[0].Shapes.GetShape(795);
            // get a sub-shape of the group shape by id
            Shape subShape = shape.Shapes.GetShape(794);

            Matrix m = new Matrix();
            // apply the translation vector
            m.Translate(-(float)subShape.XForm.LocPinX.Value, -(float)subShape.XForm.LocPinY.Value);
            // set the elements of that matrix to a rotation
            m.Rotate((float)subShape.XForm.Angle.Value);
            // apply the translation vector
            m.Translate((float)subShape.XForm.PinX.Value, (float)subShape.XForm.PinY.Value);

            // get pinx and piny
            double pinx = m.OffsetX;
            double piny = m.OffsetY;
            // calculate the sub-shape pinx and piny
            double resultx = shape.XForm.PinX.Value - shape.XForm.LocPinX.Value - pinx;
            double resulty = shape.XForm.PinY.Value - shape.XForm.LocPinY.Value - piny;
            // ExEnd:CalculateCenterOfSubShapes
        }
    }
}

File: Examples/VisualBasic/ProgrammersGuide/Working-Shapes/CalculateCenterOfSubShapes.vb
Imports VisualBasic
Imports Aspose.Diagram
Imports System
Imports System.Drawing.Drawing2D

Public Class CalculateCenterOfSubShapes
    Public Shared Sub Run()
        ' ExStart:CalculateCenterOfSubShapes
        ' The path to the documents directory.
        Dim dataDir As String = RunExamples.GetDataDir_Shapes()

        ' load Visio diagram
        Dim diagram As New Diagram(dataDir & Convert.ToString("Drawing1.vsdx"))
        ' get a group shape by ID and page index is 0
        Dim shape As Shape = diagram.Pages(0).Shapes.GetShape(795)
        ' get a sub-shape of the group shape by id
        Dim subShape As Shape = shape.Shapes.GetShape(794)

        Dim m As New Matrix()
        ' apply the translation vector
        m.Translate(-CSng(subShape.XForm.LocPinX.Value), -CSng(subShape.XForm.LocPinY.Value))
        ' set the elements of that matrix to a rotation
        m.Rotate(CSng(subShape.XForm.Angle.Value))
        ' apply the translation vector
        m.Translate(CSng(subShape.XForm.PinX.Value), CSng(subShape.XForm.PinY.Value))

        ' get pinx and piny
        Dim pinx As Double = m.OffsetX
        Dim piny As Double = m.OffsetY
        ' calculate the sub-shape pinx and piny
        Dim resultx As Double = shape.XForm.PinX.Value - shape.XForm.LocPinX.Value - pinx
        Dim resulty As Double = shape.XForm.PinY.Value - shape.XForm.LocPinY.Value - piny
        ' ExEnd:CalculateCenterOfSubShapes
    End Sub
End Class
